### PR TITLE
Fix sliver persistent header expand animation

### LINKED ISSUE #137901: SliverAppBar snap broken animation
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Add `SliverAppBar` pinned: true, floating: true, snap:true
2. Add some any `EditableText`
3.  Scroll down so appbar isn't visible
4. Tap into `EditableText` to get focus and keyboard appeared
5. Get broken animation

### Expected results

Normal animation

### Actual results

Animation is freezing or glitching

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

void main() => runApp(const NestedScrollViewExampleApp());

class NestedScrollViewExampleApp extends StatelessWidget {
  const NestedScrollViewExampleApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: NestedScrollViewExample(),
    );
  }
}

class NestedScrollViewExample extends StatefulWidget {
  const NestedScrollViewExample({super.key});

  @override
  State<NestedScrollViewExample> createState() =>
      _NestedScrollViewExampleState();
}

class _NestedScrollViewExampleState extends State<NestedScrollViewExample> {
  @override
  Widget build(BuildContext context) {
    final List<String> tabs = <String>['Tab 1', 'Tab 2'];
    return DefaultTabController(
      length: tabs.length,
      child: Scaffold(
        body: NestedScrollView(
          headerSliverBuilder: (BuildContext context, bool innerBoxIsScrolled) {
            return <Widget>[
              SliverOverlapAbsorber(
                handle:
                    NestedScrollView.sliverOverlapAbsorberHandleFor(context),
                sliver: SliverAppBar(
                  surfaceTintColor: Colors.transparent,
                  leading: IconButton(
                    onPressed: () {},
                    icon: const Icon(Icons.close),
                  ),
                  title: const Text('Books'),
                  pinned: true,
                  snap: true,
                  floating: true,
                  expandedHeight: 180.0,
                  forceElevated: innerBoxIsScrolled,
                  bottom: PreferredSize(
                    preferredSize: const Size.fromHeight(128),
                    child: Column(
                      mainAxisSize: MainAxisSize.min,
                      children: [
                        const Padding(
                          padding: EdgeInsets.symmetric(horizontal: 16.0),
                          child: TextField(
                            decoration: InputDecoration(label: Text('Input')),
                          ),
                        ),
                        TabBar(
                          // These are the widgets to put in each tab in the tab bar.
                          tabs: tabs
                              .map((String name) => Tab(text: name))
                              .toList(),
                        )
                      ],
                    ),
                  ),
                ),
              ),
            ];
          },
          body: TabBarView(
            children: tabs.map((String name) {
              return SafeArea(
                top: false,
                bottom: false,
                child: Builder(
                  builder: (BuildContext context) {
                    return CustomScrollView(
                      keyboardDismissBehavior:
                          ScrollViewKeyboardDismissBehavior.onDrag,
                      key: PageStorageKey<String>(name),
                      slivers: <Widget>[
                        SliverOverlapInjector(
                          handle:
                              NestedScrollView.sliverOverlapAbsorberHandleFor(
                                  context),
                        ),
                        SliverPadding(
                          padding: const EdgeInsets.all(8.0),
                          sliver: SliverFixedExtentList(
                            itemExtent: 48.0,
                            delegate: SliverChildBuilderDelegate(
                              (BuildContext context, int index) {
                                return ListTile(
                                  title: Text('Item $index'),
                                );
                              },
                              childCount: 30,
                            ),
                          ),
                        ),
                      ],
                    );
                  },
                ),
              );
            }).toList(),
          ),
        ),
      ),
    );
  }
}


```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

After research I've found out that the reason for such behaviour is a few things:

1. If `EditableText` is focused it will call `_scheduleShowCaretOnScreen(withAnimation: true);` that will cause animation start of a `RenderSliverFloatingPersistentHeader` (by calling `showOnScreen` with provided `Duration` of 200 milliseconds).
 image
2. After first step `EditableText` will handle software keyboard appearance and will call `_scheduleShowCaretOnScreen(withAnimation: false);` causing insta call of `showOnScreen` again but now with `Duration.zero` which leads to broken animation
image

I've tried to add dummy additional check and it fixes problem:

image

Without fix:

[untitled.webm](https://github.com/flutter/flutter/assets/72284940/7ad47dbd-15fe-40b7-b50e-ad4a8379ec4f)

With fix:

[fixed.webm](https://github.com/flutter/flutter/assets/72284940/1353009d-66f9-4e54-b43b-1eb8f271338f)

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[!] Flutter (Channel master, 3.17.0-1.0.pre.33, on macOS 13.4 22F66 darwin-arm64, locale en-RU)
    ! Upstream repository https://github.com/feduke-nukem/flutter is not a standard remote.
      Set environment variable "FLUTTER_GIT_URL" to https://github.com/feduke-nukem/flutter to dismiss this error.
[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.3)
[✓] VS Code (version 1.84.0)
[✓] Connected device (4 available)
[✓] Network resources
```

</details>

### PR DESCRIPTION
**Pull request by feduke-nukem**
Added animation status check at showOnScreen method to prevent broken animation of expanding SliverAppBar when focusing EditableText

close #137901

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/test/rendering/viewport_test.dart [hunk 1]
```diff
@@ -1314,6 +1314,46 @@ void main() {
...
+        expect(mainAxisExtent(tester, pinnedHeaderContent), lessThan(300));
+
+
+        tester.renderObject(pinnedHeaderContent).showOnScreen(
+          descendant: tester.renderObject(pinnedHeaderContent),
+          // Adding different rect to check if the second showOnScreen call
+          // leads to a different result.
+          // When the animation has forward status and the second showOnScreen
+          // is called, the new animation won't start.
+          // https://github.com/flutter/flutter/issues/137901
```
Comment (Piinks) [inline packages/flutter/test/rendering/viewport_test.dart:1341]:
```suggestion
```

File: packages/flutter/test/rendering/viewport_test.dart [hunk 2]
```diff
@@ -1314,6 +1314,46 @@ void main() {
         },
       );
 
+      testWidgetsWithLeakTracking('RenderViewportBase.showOnScreen twice almost instantly', (WidgetTester tester) async{
+        await tester.pumpWidget(
```
Comment (Piinks) [inline packages/flutter/test/rendering/viewport_test.dart:1318]:
```suggestion
        // Regression test for https://github.com/flutter/flutter/issues/137901
        await tester.pumpWidget(
```

**Review by Piinks — APPROVED** [review]
Hi @feduke-nukem, welcome. Thanks for contributing! This change LGTM.

**Comment by feduke-nukem** [comment]
@Piinks Do you think it might be a good idea to consider changing the target branch?

**Comment by Piinks** [comment]
What do you mean?

**Comment by feduke-nukem** [comment]
> What do you mean?

Is it supposed to be merged into "master" ?

**Comment by Piinks** [comment]
> Is it supposed to be merged into the "master" ? 

Yup! That is correct. We're just waiting on a second review here for this to land. I've asked for someone to take a look :)

**Comment by feduke-nukem** [comment]
Hi @Piinks.

Are there any updates?

**Review by christopherfujino — APPROVED** [review]
LGTM

**Pull request merged**